crates/attestation/src/crypto/crl.rs

```diff
@@ -0,0 +1,279 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use crate::crypto::SubjectPublicKeyInfoExt;
+
+use anyhow::{bail, Result};
+use der::Encode;
+use x509::crl::CertificateList;
+use x509::PkiPath;
+
+pub trait PkiPathCRLCheck<'a> {
+    fn check_crl(&self, crls: &'a [CertificateList<'a>]) -> Result<()>;
+}
+
+impl<'a> PkiPathCRLCheck<'a> for PkiPath<'a> {
+    fn check_crl(&self, crls: &'a [CertificateList<'a>]) -> Result<()> {
+        let mut iter = self.iter().peekable();
+        loop {
+            let cert = iter.next();
+            let next = iter.peek();
+            if cert.is_none() || next.is_none() {
+                break;
+            }
+            let cert = cert.unwrap();
+            let next = next.unwrap();
+            for crl in crls {
+                // Don't bother with an empty CRL
+                if crl.tbs_cert_list.revoked_certificates.is_none() {
+                    continue;
+                }
+
+                let revocations = crl.tbs_cert_list.revoked_certificates.as_ref().unwrap();
+                if revocations.is_empty() {
+                    continue;
+                }
+
+                // Raw bytes that are signed by the CA.
+                // Unwrap is save since we've already decoded it from der, so getting
+                // der bytes will succeed.
+                if crl.signature_algorithm != cert.signature_algorithm {
+                    // Must not be the correct Certificate-CRL match, go to the next Certificate
+                    // since the signature algorithm on the CRL should match the algorithm
+                    // type of the signing CA, right?
+                    continue;
+                }
+
+                let raw_bytes = crl.tbs_cert_list.to_vec().unwrap();
+                match cert.tbs_certificate.subject_public_key_info.verify(
+                    &raw_bytes,
+                    cert.signature_algorithm,
+                    crl.signature.raw_bytes(),
+                ) {
+                    Ok(_) => {
+                        for revoked in revocations {
+                            if revoked.serial_number == next.tbs_certificate.serial_number {
+                                bail!("revoked!");
+                            }
+                        }
+                    }
+                    Err(_) => {
+                        // Not the correct Certificate-CRL match, go to the next.
+                        continue;
+                    }
+                }
+            }
+        }
+        Ok(())
```

Comment:
This one means that if there were no valid CRLs, we just allow the certificate to go through. This is prone to an attacker mangling all applicable CRLs, after which the cert gets accepted incorrectly.

Reply:
Right, makes sense.

Reply:
Resolved, but double-check